A51-custom-backend-metrics.md

```diff
@@ -252,91 +206,119 @@ A client may change its requested interval at any time; this is expected to be a
+that the client does not need, but this case should be fairly rare;
+backends will generally not be written to populate data that they do not need.
+Also, we want to avoid the complexity of introducing some kind of negotiation for the client to
+tell the backend that it wants this data.
 
-1. Collect Utilization Metrics using Out-of-Band Reporting
+Note that reporting utilization metrics per-query is only useful in high QPS systems.
+In such systems, the application that measures utilization data should consider sampling the data
+to avoid performance degradation due to high rates of syscalls.
 
-It is popular for a gRPC powered environment to be running its own intra-locality WRR LB policy plugin 
-at the client side, which will need to access the backend utilization metrics feedback that is further 
-translated to subchannel weights for the scheduling algorithm to function, 
-e.g. CPU, queue depth. Observing these metrics is fulfilled by the gRPC custom backend metrics feature.
+For the server APIs in the OOB reporting, alternative to the set-style APIs in a push model,
+we can use callback-style APIs in a pull model to update the metrics data,
+however, this is suboptimal because it is not performance/scalability friendly.
+In a pull model, the time complexity of doing utilization measurement syscalls reaches O(M*N),
+M being the number of OOB clients and N being their frequency configuration.
+In a push model, the number of syscalls decouples with the number of OOB clients and their configurations.
+In fact, the server application is able to tune the performance by sampling the utilization data at a constant rate.
 
-2. Collect Utilization Metrics using Per-Query Reporting 
+It is by design that the OOB server sends the data regardless of whether it has been changed or 
+not since the last time it was sent. We considered a potential optimization, that is to skip 
+sending the unchanged metrics data since last response.
+To do that, we should not compare maps for de-duplication, instead, we can bump up a
+generation_id of the metrics data every time users make a change.
+Each client connection only needs to compare its previous generation_id to do deduplication.
+The optimization strategy can be implemented until there are concrete use cases in the future.
 
-In a high QPS environment, reporting request cost using per-query is desired. 
-Compared with 1, the advantage of per-query is that the utilization metrics can be timely aligned with the requests. 
-It also communicates the metrics data rapidly and saves cost of the idle connection compared with OOB reporting.
+The reporting frequency in the OOB request is final because it is a server-streaming API.
+Clients would terminate the OOB stream and reconnect if they want to change the frequency.
+Switching to a bidi stream will make it possible to change the reporting frequency in flight,
+but it should not impact the public interface.
```

Comment:
s/it should not impact/this can be done in the future if needed without impacting/